File: android-sdk/sdk/src/main/java/com/adlcom/mxl/sdk/config/RemoteConfigManager.kt
package com.adlcom.mxl.sdk.config

import android.content.Context
import android.content.SharedPreferences
import com.adlcom.mxl.sdk.internal.SdkState
import com.adlcom.mxl.sdk.internal.getSdkState
import com.adlcom.mxl.sdk.logging.Logger
import com.adlcom.mxl.sdk.network.HttpClient
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.TimeUnit

/**
 * Manages remote configuration updates.
 */
class RemoteConfigManager private constructor(
    private val context: Context,
    private val configuration: SdkConfiguration,
    private val httpClient: HttpClient,
    private val prefs: SharedPreferences,
    private val gson: Gson
) {
    companion object {
        private const val PREFS_NAME = "mxl_remote_config"
        private const val KEY_CONFIG_VERSION = "config_version"
        private const val KEY_CONFIG_DATA = "config_data"
        private const val KEY_LAST_FETCH = "last_fetch"

        fun create(context: Context, configuration: SdkConfiguration, httpClient: HttpClient): RemoteConfigManager {
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val gson = Gson()
            return RemoteConfigManager(context, configuration, httpClient, prefs, gson)
        }
    }

    /**
     * Fetch remote configuration.
     */
    fun fetchConfig() {
        val scope = CoroutineScope(Dispatchers.IO)
        scope.launch {
            try {
                val response = httpClient.post("/api/v1/config", "{}")
                
                when (response) {
                    is com.adlcom.mxl.sdk.network.NetworkResult.Success -> {
                        val configData = gson.fromJson(response.data, RemoteConfigData::class.java)
                        saveConfig(configData)
                        Logger.d("Remote config fetched: version ${configData.version}")
                    }
                    is com.adlcom.mxl.sdk.network.NetworkResult.Error -> {
                        Logger.w("Failed to fetch remote config: ${response.message}")
                    }
                }
            } catch (e: Exception) {
                Logger.e("Error fetching remote config", e)
            }
        }
    }

    /**
     * Get feature flag value.
     */
    fun getFeatureFlag(key: String, defaultValue: Boolean = false): Boolean {
        val configData = getCachedConfig() ?: return defaultValue
        return configData.featureFlags[key] ?: defaultValue
    }

    /**
     * Get sampling rate.
     */
    fun getSamplingRate(): Float {
        val configData = getCachedConfig() ?: return configuration.samplingRate
        return configData.samplingRate ?: configuration.samplingRate
    }

    /**
     * Get config value.
     */
    fun getConfigValue(key: String, defaultValue: String = ""): String {
        val configData = getCachedConfig() ?: return defaultValue
        return configData.config[key] ?: defaultValue
    }

    /**
     * Save configuration.
     */
    private fun saveConfig(configData: RemoteConfigData) {
        prefs.edit()
            .putInt(KEY_CONFIG_VERSION, configData.version)
            .putString(KEY_CONFIG_DATA, gson.toJson(configData))
            .putLong(KEY_LAST_FETCH, System.currentTimeMillis())
            .apply()
    }

    /**
     * Get cached configuration.
     */
    private fun getCachedConfig(): RemoteConfigData? {
        val configJson = prefs.getString(KEY_CONFIG_DATA, null) ?: return null
        return try {
            gson.fromJson(configJson, RemoteConfigData::class.java)
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Check if config needs refresh.
     */
    fun needsRefresh(): Boolean {
        val lastFetch = prefs.getLong(KEY_LAST_FETCH, 0)
        val now = System.currentTimeMillis()
        val timeSinceFetch = now - lastFetch
        return timeSinceFetch > TimeUnit.HOURS.toMillis(1) // Refresh every hour
    }
}

data class RemoteConfigData(
    val version: Int,
    val samplingRate: Float?,
    val featureFlags: Map<String, Boolean>,
    val config: Map<String, String>
)

File: ios-sdk/Sources/MxLSdk/Config/RemoteConfigManager.swift
import Foundation

/**
 * Manages remote configuration updates.
 */
internal class RemoteConfigManager {
    private let configuration: SdkConfiguration
    private let httpClient: HttpClient
    private let userDefaults: UserDefaults
    
    private enum Keys {
        static let configVersion = "mxl_config_version"
        static let configData = "mxl_config_data"
        static let lastFetch = "mxl_config_last_fetch"
    }
    
    static func create(configuration: SdkConfiguration, httpClient: HttpClient) -> RemoteConfigManager {
        return RemoteConfigManager(configuration: configuration, httpClient: httpClient)
    }
    
    private init(configuration: SdkConfiguration, httpClient: HttpClient) {
        self.configuration = configuration
        self.httpClient = httpClient
        self.userDefaults = UserDefaults.standard
    }
    
    /**
     * Fetch remote configuration.
     */
    func fetchConfig() async {
        do {
            let body = try JSONSerialization.data(withJSONObject: [:])
            let response = try await httpClient.post(path: "/api/v1/config", body: body)
            
            switch response {
            case .success(let data):
                if let configData = try? JSONDecoder().decode(RemoteConfigData.self, from: data) {
                    saveConfig(configData)
                    Logger.shared.info("Remote config fetched: version \(configData.version)")
                }
            case .error(let code, let message):
                Logger.shared.warning("Failed to fetch remote config: \(message)")
            }
        } catch {
            Logger.shared.error("Error fetching remote config", error: error)
        }
    }
    
    /**
     * Get feature flag value.
     */
    func getFeatureFlag(_ key: String, defaultValue: Bool = false) -> Bool {
        let configData = getCachedConfig() ?? nil
        return configData?.featureFlags[key] ?? defaultValue
    }
    
    /**
     * Get sampling rate.
     */
    func getSamplingRate() -> Float {
        let configData = getCachedConfig()
        return configData?.samplingRate ?? configuration.samplingRate
    }
    
    /**
     * Get config value.
     */
    func getConfigValue(_ key: String, defaultValue: String = "") -> String {
        let configData = getCachedConfig()
        return configData?.config[key] ?? defaultValue
    }
    
    /**
     * Save configuration.
     */
    private func saveConfig(_ configData: RemoteConfigData) {
        if let encoded = try? JSONEncoder().encode(configData) {
            userDefaults.set(configData.version, forKey: Keys.configVersion)
            userDefaults.set(encoded, forKey: Keys.configData)
            userDefaults.set(Date().timeIntervalSince1970, forKey: Keys.lastFetch)
        }
    }
    
    /**
     * Get cached configuration.
     */
    private func getCachedConfig() -> RemoteConfigData? {
        guard let data = userDefaults.data(forKey: Keys.configData) else {
            return nil
        }
        
        return try? JSONDecoder().decode(RemoteConfigData.self, from: data)
    }
    
    /**
     * Check if config needs refresh.
     */
    func needsRefresh() -> Bool {
        let lastFetch = userDefaults.double(forKey: Keys.lastFetch)
        let timeSinceFetch = Date().timeIntervalSince1970 - lastFetch
        return timeSinceFetch > 3600 // 1 hour
    }
}

struct RemoteConfigData: Codable {
    let version: Int
    let samplingRate: Float?
    let featureFlags: [String: Bool]
    let config: [String: String]
}

